src/combinatorics/rado/rank_function.lean

```diff
@@ -0,0 +1,235 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+
+import combinatorics.rado.auxiliary
+
+/-!
+# Rank functions
+
+We define the type of rank functions on a type `α` and prove some properties.
+
+See the docstring of `rank_fn` for the definition.
+-/
+
+section rk_fn_def
+
+variables (α : Type*) [decidable_eq α]
+
+/-- A *rank function* on a type `α` is a function `r` from `finset α` to natural numbers such that:
+* `r ∅ = 0`
+* `r s ≤ r ({a} ∪ s)`
+* `r ({a} ∪ s) ≤ r s + 1`,
+* `r ({a} ∪ s) = r s` and `r ({b} ∪ s) = r s` imply `r ({a, b} ∪ s) = r s`
+
+(This definition is equivalent to the definition one can find on the
+(Wikipedia page for matroids)[https://en.wikipedia.org/wiki/Matroid].
+But note that we do not assume that the base type `α` is finite.)
+-/
+@[ext]
+structure rank_fn :=
+(to_fun : finset α → ℕ)
+(empty' : to_fun ∅ = 0)
+(le_insert' : ∀ a s, to_fun s ≤ to_fun (insert a s))
+(insert_le' : ∀ a s, to_fun (insert a s) ≤ to_fun s + 1)
+(dep' : ∀ {s a b}, to_fun (insert a s) = to_fun s → to_fun (insert b s) = to_fun s →
+          to_fun (insert b (insert a s)) = to_fun s)
```

Comment:
I wonder whether it is useful to generalise this from `finset alpha` to any graded order.

Reply:
I guess the question is what you would want to do with it...

Reply:
There is a bit of a mathematical issue with this definition. It is true that in the finite case, rank functions as you've defined them describe matroids, but this isn't true for infinite matroids. In particular, there are many distinct matroids whose finite sets all have rank equal to their cardinality. For instance, I can take the ground set to be some infinite `E`, declare every set except `E` itself to be independent, and then either choose `E` to be independent or dependent - the rank function on finite sets can't detect the difference between these two matroids. 

The path to axiomatizing infinite matroids correctly was a little fraught, with various authors trying things like rank functions on finite sets, before a single correct notion was found - see https://www.math.uni-hamburg.de/home/diestel/papers/matroidaxioms.pdf  for the appropriate definitions and historical discussion. Rank functions, even when they are allowed to be defined on infinite sets, turn out to be insufficient - one needs a 'relative rank' function for pairs of sets. 

For this reason, I'd be a little worried about matroids being introduced to mathlib in the way you've defined them - in short 'rank functions satisfying matroid axioms whose domain is the collection of finite sets in `E`' isn't a natural class of mathematical objects unless `E` is finite. To point out a concrete problem, there is no reasonable theory of infinite matroid duality for your definition. 

My matroid repository  ( https://github.com/apnelson1/lean-matroids ) defines matroids in a way that follows the standard notion of infinite matroids; I am only waiting on https://github.com/leanprover-community/mathlib/pull/18576 for my first PR. I think you should be able to adapt your proof of Rado to that setting.

Reply:
I'm not claiming to implement infinite matroids here; I just wanted something that satisfies a kind of minimal set of conditions so that I can prove Rado's Theorem (and apply it to get Halls' Theorem and the linear algebra version). Would the more general kind of rank function be as easy to work with?

I see one potential problem with switching to a notion compatible with infinite matroids (disclaimer: I know nothing about infinite matroids (yet)), looking at what you write, and this is that the version for arbitrary indexing types would be wrong in general. If I take your example where `E` is infinite and all finite subsets are independent, but `E` is dependent, I could take the indexing type to be `E` and consider the family of singleton sets, `Z i = {i}`. This satisfies the condition (for finite subfamilies), but the only "section" there is would not be independent. So there is perhaps a question how to reconcile general infinite matroids and the version of the theorem as proved here. I assume the theorem would apply to infinite matroids satisfying the condition that a set is independent if all its finite subsets are.

Reply:
Yes, finitary matroids are a natural class encompassing your setting, and they seem like a reasonable setting for Rado to hold. For nonfinitary matroids, things close to Rado are still open questions. 

Looking at your proof, it seems like it should adapt quite easily to this class. I realise you weren't claiming to do infinite matroids completely with your definition; my point was just that it is not the right way to add matroids to mathlib. 

I think my matroids project should have all the rank function API you use, and I had been planning to add `[finitary M]` as a typeclass soon. If you'd be interested in refactoring it to a general matroid setting, I can give you access to the project.

Reply:
I am in no hurry with this. It probably makes sense in any case to wait for the port to include the files that this depends on and then do this for mathlib4 (hopefully there will be an easy way to translate the code by then). Also, teaching has just started again for me, so my bandwidth is a bit limited...

I do agree that it should make sense to do the proof based on a suitable API for rank functions etc., to avoid unnecessary duplication. For the intended use (which was my motivation for formalizing this proof and which does not aim at mathlib), I can use what I have locally in any case.